### PR TITLE
Close streams when a connection is terminated

### PR DESCRIPTION
**Pull request by snorp**
This also adds some things to let node-http2 emit the
proper events.

### CODE REVIEW

**Comment by nwgh** [comment]
Ugh... nevermind my previous comment, I was totally misunderstanding what was going on. Blame me being exhausted :)

File: lib/connection.js [hunk 1]
```diff
@@ -356,14 +359,25 @@ Connection.prototype._receive = function _receive(frame, done) {
   // * gets the appropriate stream from the stream registry
   var stream = this._streamIds[frame.stream];
 
-  // * or creates one if it's not in `this.streams`
   if (!stream) {
+    // This is a PROTOCOL_ERROR if it's not HEADERS and or PUSH_PROMISE.
```
Comment (molnarg) [inline lib/connection.js]: This situation is handled by the stream state machine, don't need to check this here.

**Comment by molnarg** [comment]
So there is two separate situation here and they are mixed in the implementation. The public `close()` method should be the 'public interface' for GOAWAY. This is what the spec says about GOAWAY:

```
   The purpose of
   this frame is to allow an endpoint to gracefully stop accepting new
   streams (perhaps for a reboot or maintenance), while still finishing
   processing of previously established streams.
```

If we write reset to all the streams, they won't finish gracefully for sure :)

The other situation when the other side closes the TCP stream. Handling this should ho into a separate function, and that function could do what you implemented in the `close()` method.

**Comment by snorp** [comment]
Yeah, good points here. I added Connection.notifySocketClosed(), which does as you say. I am not so sure about calling 'close' or 'notifySocketClosed' in response to the 'finish' event now. I think we probably want to call notifySocketClosed() in response to 'finish', but have an early return there if 'this._closed === true'. Calling Connection.close() and the pipe calling end() will both trigger a 'finish'. Only the latter should forcibly close the streams, but I may be missing something.

**Comment by snorp** [comment]
Updated to fix the review comments

**Comment by molnarg** [comment]
I've started merging this, but I've realised that the current GOAWAY handling was in a terrible state in node-http2-protocol, so ended up rewriting a large part in the end. What I copying almost without change is the invalid stream ID handling and the default error code in stream reset().

This is what I've came up with:

https://github.com/molnarg/node-http2-protocol/compare/6121463...master

With this change, I believe node-http2 patch can be somewhat smaller but didn't work on that yet, will have a look at it soon. Sorry that I did not follow the usual review cycle workflow here, but this needed some refactoring and felt that it will be quicker and simpler this way.

**Comment by molnarg** [comment]
Just to clarify: what I realised is that the GOAWAY handling in node-http2-protocol (and not in your patch) is basically messed up and needs rewrite.

**Comment by nwgh** [comment]
Ugh, so Gábor, your commits totally broke node-http2 with master. Unless I hear back, I think I'd like to back them out (so we have master working with master in the two repos), and then we can do work in branches to get them working together (or, better yet, I can just get off my butt and merge the 2 repos again.)

**Comment by molnarg** [comment]
Um, sorry for breaking node-http2. So here is my plan: today I'll try update node-http2 to fix things, and in case I won't manage to do this, we can back this change out and do something like this after the merge.

**Comment by nwgh** [comment]
Sounds good to me :) I'd have tried to fix it myself, but I've been too busy "breaking twitter" and other fun things!

**Comment by molnarg** [comment]
I've just reverted the recent changes! I've run into some tricky issue with proper handling of the connection end, and will continue working on it in the close-streams-2 branch.

**Pull request closed without merging**